test/unit/cdylib_name/cdylib_name_analysis_test.bzl

```
@@ -0,0 +1,46 @@
+"""Analysis tests for the name we assign to cdylib libraries."""
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts")
+load("//rust:defs.bzl", "rust_shared_library")
+
+def _cdylib_name_test_impl(ctx):
+    env = analysistest.begin(ctx)
+    target = analysistest.target_under_test(env)
+
+    # We're expecting the `.dylib`/`.so` to be the only file on Unix and Windows to
+    # contain a pair of `.dll` and `.lib` files.
+    files = target[DefaultInfo].files.to_list()
+    if len(files) == 1:
+        asserts.true(env, files[0].extension in ("so", "dylib"))
+        if files[0].extension == "so":
+            asserts.equals(env, files[0].basename, "libsomething.so")
+        elif files[0].extension == "dylib":
+            asserts.equals(env, files[0].basename, "libsomething.dylib")
+    elif len(files) == 2:
+        asserts.equals(env, files[0].basename, "something.dll")
```

This might be a minor thing but I'd like to avoid creating a dependency on the ordering of the files returned by `DefaultInfo`. I think everyone should do due-diligence to ensure the files they're trying to get out of a list is what they want. Depending on `depset().to_list()[INDEX]` is something I'm hoping we can avoid
```suggestion
        expected_extensions = ["dll", "lib"]
        for file in files:
            asserts.true(env, file.extension in expected_extensions)
            expected_extensions.pop(file.extension)
```

---

oh right that's fair. I'll change it to `expected_filenames` so that we can actually test the cdylib name too in one pass.

---

Sorry, my understanding of `pop` seems to be wrong. I think I meant [remove](https://docs.bazel.build/versions/4.0.0/skylark/lib/list.html#remove)

---

oops, fixed (that'll teach me not to test my windows changes on linux, I guess that's why we have CI 😄)